src/app/layout/qgsgeopdflayertreemodel.cpp

```diff
@@ -0,0 +1,170 @@
+{
+  if ( idx.column() == LayerColumn )
+  {
+    return QgsLayerTreeModel::flags( idx ) | Qt::ItemIsUserCheckable;
+  }
+
+  QgsVectorLayer *vl = vectorLayer( idx );
+  if ( !vl )
+  {
+    return Qt::NoItemFlags;
+  }
+  else
+  {
+    const QModelIndex layerIndex = sibling( idx.row(), LayerColumn, idx );
+    if ( data( layerIndex, Qt::CheckStateRole ) == Qt::Checked )
+    {
+      return Qt::ItemIsEnabled | Qt::ItemIsEditable;
+    }
+  }
+  return Qt::NoItemFlags;
+}
+
+QgsVectorLayer *QgsGeoPdfLayerTreeModel::vectorLayer( const QModelIndex &idx ) const
+{
+  QgsLayerTreeNode *node = index2node( index( idx.row(), LayerColumn, idx.parent() ) );
+  if ( !node || !QgsLayerTree::isLayer( node ) )
+    return nullptr;
+
+  return qobject_cast<QgsVectorLayer *>( QgsLayerTree::toLayer( node )->layer() );
+}
+
+QVariant QgsGeoPdfLayerTreeModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  if ( orientation == Qt::Horizontal )
+  {
+    if ( role == Qt::DisplayRole )
+    {
+      switch ( section )
+      {
+        case 0:
+          return tr( "Layer" );
+        case 1:
+          return tr( "PDF Group" );
+        default:
+          return QVariant();
+      }
+    }
+  }
+  return QgsLayerTreeModel::headerData( section, orientation, role );
+}
+
+QVariant QgsGeoPdfLayerTreeModel::data( const QModelIndex &idx, int role ) const
+{
+  switch ( idx.column() )
+  {
+    case LayerColumn:
+    {
+      if ( role == Qt::CheckStateRole )
+      {
+        QgsLayerTreeNode *node = index2node( index( idx.row(), LayerColumn, idx.parent() ) );
+        QgsVectorLayer *vl = vectorLayer( idx );
+        if ( vl )
+        {
+          const QVariant v = vl->customProperty( QStringLiteral( "geopdf/includeFeatures" ) );
+          if ( v.isValid() )
+          {
+            return v.toBool() ? Qt::Checked : Qt::Unchecked;
+          }
+          else
+          {
+            // otherwise, we default to the layer's visibility
+            return node->itemVisibilityChecked() ? Qt::Checked : Qt::Unchecked;
+          }
+        }
+        return QVariant();
+      }
+      return QgsLayerTreeModel::data( idx, role );
+    }
+    case GroupColumn:
+    {
+      switch ( role )
+      {
+        case Qt::DisplayRole:
+        {
+          if ( QgsVectorLayer *vl = vectorLayer( idx ) )
+          {
+            return vl->customProperty( QStringLiteral( "geopdf/groupName" ) ).toString();
+          }
+          break;
+        }
+      }
+
+      return QVariant();
+    }
+  }
+
+  return QVariant();
+}
+
+bool QgsGeoPdfLayerTreeModel::setData( const QModelIndex &index, const QVariant &value, int role )
+{
+  switch ( index.column() )
+  {
+    case LayerColumn:
+    {
+      if ( role == Qt::CheckStateRole )
+      {
+        if ( QgsVectorLayer *vl = vectorLayer( index ) )
+        {
+          vl->setCustomProperty( QStringLiteral( "geopdf/includeFeatures" ), value.toInt() == Qt::Checked );
```

Comment:
You've decided to attach those flags to the vector layer, which means it'll be a setting that will be carried across layouts. Is there a reason to think this would ever become problematic / undesirable. I.e., would those settings better sit as custom properties of a layout?

Reply:
I debated this too. Ultimately whichever way we go we'll frustrate one set of users! I think it's more likely that settings will apply across a whole project though, which is why I've gone with this approach.

Reply:
@nyalldawson , I can see the dilemma. I trust your gut then.